### PR TITLE
Small improvements

### PR DESCRIPTION
**Pull request by kingcrimsontianyu**
This PR makes a couple of small, general improvements.
- Improve the use of `std::tolower` to avoid UB. References [cppreference](https://en.cppreference.com/w/cpp/string/byte/tolower).
- Improve the readability of POSIX read/write interface.
- Improve the use of `std::map` to avoid a repeated search in the self-balancing BST, which takes `O(n)`.

### CODE REVIEW

File: cpp/include/kvikio/posix_io.hpp [hunk 1]
```diff
@@ -183,18 +206,20 @@ std::size_t posix_device_io(int fd,
...
+ * @tparam ioDataCompletionLevel Whether all requested data are processed or not. If `FULL`, all of
+ * `count` bytes are read.
  * @param fd File descriptor
  * @param buf Base address of buffer in host memory.
  * @param size Size in bytes to read.
  * @param file_offset Offset in the file to read from.
- * @param partial If false, all of `size` bytes are read.
  * @return Size of bytes that were successfully read.
  */
-inline std::size_t posix_host_read(
```
Comment (kingcrimsontianyu) [inline cpp/include/kvikio/posix_io.hpp:193]: This change and the one to write are breaking. Is this acceptable? @madsbk

File: cpp/include/kvikio/defaults.hpp [hunk 2]
```diff
@@ -57,7 +57,9 @@ inline bool getenv_or(std::string_view env_var_name, bool default_val)
   }
   // Convert to lowercase
   std::string str{env_val};
-  std::transform(str.begin(), str.end(), str.begin(), ::tolower);
+  // To avoid UB in std::tolower(), the character must be cast to unsigned char
```
Comment (mythrocks) [inline cpp/include/kvikio/defaults.hpp:60]: 😮

File: cpp/include/kvikio/posix_io.hpp [hunk 3]
```diff
@@ -84,43 +104,44 @@ class StreamsByThread {
...
  * @param fd File descriptor
  * @param buf Buffer to write
  * @param count Number of bytes to write
  * @param offset File offset
- * @param partial If false, all of `count` bytes are read or written.
  * @return The number of bytes read or written (always gather than zero)
  */
-template <bool IsReadOperation>
-ssize_t posix_host_io(int fd, const void* buf, size_t count, off_t offset, bool partial)
+template <IOType ioType, IODataCompletionLevel ioDataCompletionLevel>
```
Comment (wence-) [inline cpp/include/kvikio/posix_io.hpp:116]: Is there really a benefit to templating on `IODataCompletionLevel`? It adds a single break to the loop which should be predicted correctly every time by the branch predictor.

File: cpp/include/kvikio/posix_io.hpp [hunk 4]
```diff
@@ -28,6 +28,24 @@
...
+
+/**
+ * @brief The degree to which the requested size of data is completed.
+ *
+ */
+enum class IODataCompletionLevel : uint8_t {
+  PARTIAL,  ///< POSIX read/write is called only once, which may not process all data requested.
+  FULL,     ///< POSIX read/write is called repeatedly until all requested data are processed.
+};
+
```
Comment (wence-) [inline cpp/include/kvikio/posix_io.hpp:48]: I wonder if, rather than having two enums, it makes more sense to spell out the Cartesian product in the `IOType` enum:

```
enum class IOType : uint8_t {
   READ_PARTIAL,
   READ_FULL,
   WRITE_PARTIAL,
   WRITE_FULL,
};
```

If not, I'm not sure I love `IODataCompletionLevel` as a name. How about:

- `PartialIO::YES/NO`
- `IsPartial::YES/NO`

?

File: cpp/include/kvikio/posix_io.hpp [hunk 5]
```diff
@@ -131,15 +152,15 @@ ssize_t posix_host_io(int fd, const void* buf, size_t count, off_t offset, bool
...
+ * @tparam ioType Whether the operation is a read or a write.
  * @param fd File descriptor
  * @param devPtr_base Device pointer to read or write to.
  * @param size Number of bytes to read or write.
  * @param file_offset Byte offset to the start of the file.
  * @param devPtr_offset Byte offset to the start of the device pointer.
  * @return Number of bytes read or written.
  */
-template <bool IsReadOperation>
+template <IOType ioType>
```
Comment (wence-) [inline cpp/include/kvikio/posix_io.hpp:163]: *nit*: I think one prefers template parameters to start with an uppercase letter. I think since every usage of the value is matched up with comparison with an `IOType::XXX` then we can get away with not writing `io` in the name as well.
```suggestion
template <IOType Type>
```

**Review by wence- — COMMENTED** [review]
A few stylistic suggestions.

File: cpp/include/kvikio/posix_io.hpp (hunk 1 above)
Comment (mythrocks) [reply cpp/include/kvikio/posix_io.hpp:193]: Watching this, to see how breaking changes are best handled.  (One option would have been to deprecate in one release, and then remove in the other.)

File: cpp/include/kvikio/posix_io.hpp (hunk 4 above)
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp:48]: Thanks for the suggestion! The first 4-choice enum looks very succinct. I'll make the changes.

File: cpp/include/kvikio/posix_io.hpp (hunk 3 above)
Comment (mythrocks) [reply cpp/include/kvikio/posix_io.hpp:116]: > which should be predicted correctly every time by the branch predictor.

@wence-: ELI5 how we know that that would be predicted correctly?  (Asking out of ignorance.)

Would that not also apply to `IsReadOperation` in the prior/original version?

File: cpp/include/kvikio/posix_io.hpp (hunk 4 above)
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp:48]: On a second thought, I probably should keep the two enums and rename the 2nd one. There's some case like [this](https://github.com/rapidsai/kvikio/pull/493/files#diff-ce16f202f49a0591c9c773df9a234a0f2198885e23ffb29d49e1e466f4d9bcedR267) where callers only need to specify read/write, and not partial/full.

File: cpp/include/kvikio/posix_io.hpp (hunk 3 above)
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp:116]: My intention was to make the interface more consistent. Either `IsReadOperation` and  `partial` are both template parameters or both function parameters.

File: cpp/include/kvikio/posix_io.hpp [hunk 6]
```diff
@@ -28,6 +28,24 @@
 
 namespace kvikio {
 
+/**
+ * @brief Type of the IO operation.
+ *
```
[the remark this answers is not in the document]
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp]: Completed.

File: cpp/include/kvikio/posix_io.hpp [hunk 7]
```diff
@@ -28,6 +28,24 @@
...
+ */
+enum class IOType : uint8_t {
+  READ,   ///< POSIX read.
+  WRITE,  ///< POSIX write.
+};
+
+/**
+ * @brief The degree to which the requested size of data is completed.
+ *
```
[the remark this answers is not in the document]
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp]: Completed.

File: cpp/include/kvikio/posix_io.hpp (hunk 4 above)
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp]: Completed.

File: cpp/include/kvikio/posix_io.hpp (hunk 5 above)
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp]: Completed.

File: cpp/include/kvikio/defaults.hpp (hunk 2 above)
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/defaults.hpp]: I've expanded the initial comment, since there's actually another issue on top of UB: `std::tolower` is not an addressable function, and passing it as a function pointer to the STL algorithm functions would cause the program behavior to be "[unspecified (possibly ill-formed)](https://eel.is/c++draft/namespace.constraints#namespace.std-6)". C++ is wild.

Comment (mythrocks) [reply cpp/include/kvikio/defaults.hpp]: > `std::tolower` is not an addressable function, and passing it as a function pointer to the STL algorithm functions would cause the program behavior to be "[unspecified (possibly ill-formed)](https://eel.is/c++draft/namespace.constraints#namespace.std-6)". C++ is wild.

What in the (hello) world...!

File: cpp/include/kvikio/defaults.hpp [hunk 8]
```diff
@@ -57,7 +57,13 @@ inline bool getenv_or(std::string_view env_var_name, bool default_val)
...
   // Convert to lowercase
   std::string str{env_val};
-  std::transform(str.begin(), str.end(), str.begin(), ::tolower);
+  // Special considerations regarding the case conversion:
+  // - std::tolower() is not an addressable function. Passing it to std::transform() as
+  //   a function pointer causes the program behavior "unspecified (possibly ill-formed)",
+  //   hence the lambda.
+  // - To avoid UB in std::tolower(), the character must be cast to unsigned char.
+  std::transform(
+    str.begin(), str.end(), str.begin(), [](unsigned char c) { return std::tolower(c); });
```
Comment (madsbk) [inline cpp/include/kvikio/defaults.hpp:66]: wow, good catch!

File: cpp/include/kvikio/posix_io.hpp (hunk 1 above)
Comment (madsbk) [reply cpp/include/kvikio/posix_io.hpp:193]: I agree, a deprecation cycle is a good thing however, it is unlikely that anyone is calling this function directly. At least, a search on github [doesn't show anything](https://github.com/search?q=posix_host_read&type=code).

In this case, I think we should allow a breaking chance without deprecation. In fact, I think we should put all of the functions in `posix_io.hpp` under `kvikio::detail`, they are not meant to be used directly.

File: cpp/include/kvikio/posix_io.hpp [hunk 9]
```diff
@@ -28,6 +28,22 @@
...
+};
+
+/**
+ * @brief Specifies whether all requested bytes are to be processed or not.
+ */
+enum class PartialIO : uint8_t {
+  YES,  ///< POSIX read/write is called only once, which may not process all bytes requested.
+  NO,   ///< POSIX read/write is called repeatedly until all requested bytes are processed.
+};
+
```
Comment (madsbk) [inline cpp/include/kvikio/posix_io.hpp:46]: Move this inside `detail` as well?

Comment (madsbk) [reply cpp/include/kvikio/posix_io.hpp:46]: i.e. just use 
```c++
namespace kvikio::detail {
```

Comment (kingcrimsontianyu) [reply cpp/include/kvikio/posix_io.hpp]: Good idea! Completed.

**Review by madsbk — APPROVED** [review]
Great work @kingcrimsontianyu

File: cpp/include/kvikio/defaults.hpp (hunk 8 above)
Comment (kingcrimsontianyu) [reply cpp/include/kvikio/defaults.hpp:66]: Is there an advantage of using `::tolower` over `std::tolower`? From the references, they seem functionally the same. I guess perhaps the reason is passing `::tolower` would make the initial code compile, while `std::tolower` doesn't (instead the template version `std::tolower(int, const std::locale&)` is picked, which doesn't compile due to two parameters).

File: cpp/include/kvikio/defaults.hpp [hunk 10]
```diff
@@ -57,7 +57,13 @@ inline bool getenv_or(std::string_view env_var_name, bool default_val)
   }
   // Convert to lowercase
   std::string str{env_val};
-  std::transform(str.begin(), str.end(), str.begin(), ::tolower);
+  // Special considerations regarding the case conversion:
+  // - std::tolower() is not an addressable function. Passing it to std::transform() as
+  //   a function pointer causes the program behavior "unspecified (possibly ill-formed)",
```
Comment (vuule) [inline cpp/include/kvikio/defaults.hpp:62]: does this point need to be removed? based on https://github.com/rapidsai/kvikio/pull/493/files#r1795596566

Comment (wence-) [reply cpp/include/kvikio/defaults.hpp:62]: We are using std::tolower in this PR, so I think this comment is correct.

Comment (kingcrimsontianyu) [reply cpp/include/kvikio/defaults.hpp]: I've expanded the comment a bit more to clarify.

**Review by mythrocks — APPROVED** [review]
Good work, this.
